nautobot/users/tests/test_api.py

```diff
@@ -73,6 +73,173 @@ def setUpTestData(cls):
+        obj_perm.save()
+        obj_perm.users.add(self.basic_auth_user_granted)
+        obj_perm.object_types.add(ContentType.objects.get_for_model(self.model))
+
+        self.basic_auth_user_permissionless = User.objects.create_user(
+            username="basicuserpermissionless", password=self.basic_auth_user_password
+        )
+
+        self.create_data = [
+            {
+                "description": "token1",
+            },
+            {
+                "description": "token2",
+            },
+            {
+                "description": "token3",
+            },
+        ]
+
+    def test_create_token_basic_authentication(self):
+        """
+        Test the provisioning of a new REST API token given a valid username and password.
+        """
+        self.client.login(username=self.basic_auth_user_granted.username, password=self.basic_auth_user_password)
+        response = self.client.post(self._get_list_url())
+        self.client.logout()
+
+        self.assertEqual(response.status_code, 201)
+        self.assertIn("key", response.data)
+        self.assertEqual(len(response.data["key"]), 40)
+        token = Token.objects.get(user=self.basic_auth_user_granted)
+        self.assertEqual(token.key, response.data["key"])
+
+    def test_create_token_basic_authentication_permissionless_user(self):
+        """
+        Test the behavior of the token create view when a user cannot create tokens
+        """
+        self.client.login(username=self.basic_auth_user_permissionless.username, password=self.basic_auth_user_password)
+        response = self.client.post(self._get_list_url())
+        self.client.logout()
+
+        self.assertEqual(response.status_code, 403)
+
+    def test_create_token_basic_authentication_invalid_password(self):
+        """
+        Test the behavior of the token create view when an invalid password is supplied
+        """
+        self.client.login(username=self.basic_auth_user_granted.username, password="hunter2")
+        response = self.client.post(self._get_list_url())
+        self.client.logout()
+
+        self.assertEqual(response.status_code, 403)
+
+    def test_create_token_basic_authentication_invalid_user(self):
+        """
+        Test the behavior of the token create view when the user supplied is not a valid user
+        """
+        self.client.login(username="iamnotreal", password="P1n0cc#10")
+        response = self.client.post(self._get_list_url())
+        self.client.logout()
+
+        self.assertEqual(response.status_code, 403)
+
+    def test_create_other_user_token_restriction(self):
+
+        # List all tokens available to user1
+        self.add_permissions("users.add_token")
+        self.add_permissions("users.change_token")
+        self.add_permissions("users.view_token")
+        previous_token_count = len(Token.objects.filter(user=self.basic_auth_user_granted))
```

Comment:
This will be zero, won't it? The only Tokens that are getting created in `setUp` are for `self.user`, not `self.basic_auth_user_granted`.

Reply:
Sure, this is meant to be flexible if for some reason it begins non-zero. The test is to ensure that following the next line the count of tokens did not go up by one (being able to issue a token for another user).